### DISCUSSION TITLE
Windows fails applying themed resources

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno#18081] [Q&A] [answered]
Applying a ThemeDictionay fails in windows but it works in desktop and wasm.

It seems that `SystemThemeHelper.SetApplicationTheme(McWindowEx.RootFrame.XamlRoot, ElementTheme.Light);` doesn't change `Application.Current.RequestedTheme`, but if I try to change it when I use the ThemeHelper, then I receive an unexpected exception when I try to use the Brush even if it has a value.

```xml
<ResourceDictionary xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">

	<SolidColorBrush x:Key="KeyboardBackgroundDark" Color="#FF000000" />
	<SolidColorBrush x:Key="KeyboardBackgroundLight" Color="#f2f2f2" />

	<ResourceDictionary.ThemeDictionaries>
		<ResourceDictionary x:Key="Light">
			<StaticResource x:Key="KeyboardBackground" ResourceKey="KeyboardBackgroundLight" />
		</ResourceDictionary>
		<ResourceDictionary x:Key="Dark">
			<StaticResource x:Key="KeyboardBackground" ResourceKey="KeyboardBackgroundDark" />
		</ResourceDictionary>
	</ResourceDictionary.ThemeDictionaries>

</ResourceDictionary>
```

```csharp
public sealed partial class KeyboardControl : Panel
{
    public KeyboardControl()
    {
        LoadResources();
        ActualThemeChanged += (s, e) => { ApplyThemedResources(); };
        Loaded += (s, e) => { ApplyThemedResources(); };
    }

    private void LoadResources()
    {
        var resources = new ResourceDictionary() { Source = new Uri("ms-appx:///UnoKeyboard/Keyboard.xaml") };
        Application.Current.Resources.MergedDictionaries.Add(resources);
    }

    private void ApplyThemedResources()
    {
        Background = (Brush)Application.Current.Resources["KeyboardBackground"];
    }
```

**Comment by MartinZikmund** [comment]
@mcNets Indeed, the name of the helper is a bit confusing, as it actually only changes the theme of the root element of the application:

https://github.com/unoplatform/uno.toolkit.ui/blob/934914f5fd65f6a1edcb1b9e0389c0d26bc4a993/src/Uno.Toolkit.UI/Helpers/SystemThemeHelper.cs#L76-L78

In case of WinUI, it is not possible to change the `Application.RequestedTheme` after app is started - you can only change the theme of the root of your visual tree instead

**Reply by mcNets** [reply]
Ok @MartinZikmund , but -desktop and -wasm projects get the proper values from the ThemedDictionary. Are they using another approach?

And how can I get the current theme within a XAML controls library?

https://github.com/unoplatform/uno.toolkit.ui/blob/934914f5fd65f6a1edcb1b9e0389c0d26bc4a993/src/Uno.Toolkit.UI/Helpers/SystemThemeHelper.cs#L25C3-L32C4

Just because background is Black?

**Reply by MartinZikmund** [reply]
The `UISettings` are retrieving the color values directly from the system - so that is the actual "OS" theme, even if the `Application.RequestedTheme` was explicitly set during app startup.

The difference in behavior in -desktop and -wasm is curious, maybe this can be reported as an issue to investigate

**Reply by mcNets** [reply]
If you want to have a look, you can run the UnoKeyboardDemo, it's using WinUI recources and it shows this behavior.

https://github.com/mcNets/UnoKeyboard/blob/3f64795fc2547dace5fa8f6173734987cf44858a/src/UnoKeyboard/Controls/KeyboardControl.cs#L21-L28

**Reply by mcNets** [reply, marked as answer]
@MartinZikmund It seems something related to WinUI, so I opened a discussion there.

https://github.com/microsoft/microsoft-ui-xaml/discussions/9943
